### PR TITLE
Accordion

### PR DESCRIPTION
**Pull request by roblevintennis**
**Description:**

- [Asana Task](https://app.asana.com/0/1150068311310825/1162798638929451/f)
- EDS Accordian

**Is this a new component? Please review these reminders: **

- [x] Have exported the component from the main [index.js](https://github.com/getethos/ethos-design-system/blob/master/src/components/index.js)?
- [x] Have you exported it from the [index.d.ts](https://github.com/getethos/ethos-design-system/blob/master/src/components/index.d.ts) so Typescript users can consume it? Added a test and ran `yarn test:types`

### CODE REVIEW

File: src/components/Accordion/AccordionSection.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+  return (
+    <>
+      <div
+        role="button"
+        aria-expanded={expanded}
+        aria-controls={sectionId}
+        id={labelId}
+        tabIndex={0}
+        className={styles.Label}
```
Comment (stephaniel96) [inline src/components/Accordion/AccordionSection.js]: I'm thinking about what needs we have for `Nora` and was wondering if we could make the style of the `title` customizable? Right now, Nora section headers have pretty particular styling

image

File: src/components/Accordion/AccordionSection.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        onClick={onClick}
+        onKeyDown={onKeyDown}
+        onFocus={onFocus}
+        onBlur={onBlur}
+        ref={labelRef}
+      >
+        {title}
```
Comment (stephaniel96) [inline src/components/Accordion/AccordionSection.js]: I'm thinking about what needs we have for `Nora` and was wondering if we could make the style of the `title` customizable? Right now, Nora section headers have pretty particular styling

image

Comment (roblevintennis) [reply src/components/Accordion/AccordionSection.js]: so you just want to be able to pass in a `className` right?

Comment (stephaniel96) [reply src/components/Accordion/AccordionSection.js]: Yeah I think that would work! I also wonder if we should separate `title` and `renderToggle(expanded)`. In the mocks right now, it seems to me that the click target is only the "+Add" text and not the entire row. Not sure what the ideal behavior should be though

**Comment by roblevintennis** [comment]
We now have a second example in the Accordion page that shows how to use `toggleChildIsTarget` boolean prop to indicate that we should only toggle sections off the toggle (not the entire header bar):

image

image

File: src/components/index.tests.tsx [hunk 3]
```diff
@@ -324,10 +339,6 @@ class ZipInputTest extends React.Component<any, any> {
   }
 }
 
-///////////////////////////////
```
Comment (roblevintennis) [inline src/components/index.tests.tsx]: We're not keeping this ordering so I figured it'd be better to just remove this as it's not become misleading

File: src/components/index.d.ts [hunk 4]
```diff
@@ -1037,6 +1037,39 @@ export declare const NoraDrawer: {
...
+  ({
+    index,
+    title,
+    renderToggle,
+    children,
+  }: {
+    index?: number
+    title: string
+    renderToggle?: any
+    children?: React.ReactNode
```
Comment (stephaniel96) [inline src/components/index.d.ts]: In the component file, `children` seems to be a required `prop`.  I think it makes sense for `children` to be required?

File: PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -11,6 +11,7 @@ _Please pick one and delete options that are not relevant:_
 - [ ] Have exported the component from the main [index.js](https://github.com/getethos/ethos-design-system/blob/master/src/components/index.js)?
 - [ ] Have you exported it from the [index.d.ts](https://github.com/getethos/ethos-design-system/blob/master/src/components/index.d.ts) so Typescript users can consume it? Added a test and ran `yarn test:types`?
 - [ ] Followed the checklist at the bottom of the [contributing guidelines](https://eds.ethoslabs.io/#/Guidelines?id=section-contribute) guide?
+- [ ] Bumped the yarn version?
```
Comment (stephaniel96) [inline PULL_REQUEST_TEMPLATE.md]: Thanks!

File: src/components/index.d.ts [hunk 6]
```diff
@@ -1037,6 +1037,39 @@ export declare const NoraDrawer: {
...
+  ({
+    id,
+    expanded,
+    onToggle,
+    children,
+  }: {
+    id?: string
+    expanded: object
+    onToggle: any
+    children: React.ReactNode
```
Comment (stephaniel96) [inline src/components/index.d.ts]: I think the `Accordion` props need to be updated with the new `toggleChildIsTarget` prop

File: src/components/Accordion/Accordion.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  onToggle: PropTypes.func.isRequired,
+  /** The Accordion's children. Likely AccordionSection's */
+  children: PropTypes.node.isRequired,
+  /** Boolean prop that indicates that you'd like to use that as the toggle (not the entire
+   * `AccordionSection` header bar). Defaults to `false` */
+  toggleChildIsTarget: PropTypes.bool,
+}
+
+Accordion.defaultProps = {
+  expanded: null,
```
Comment (stephaniel96) [inline src/components/Accordion/Accordion.js]: if `toggleChildIsTarget` defaults to `false`, should we have it in the `defaultProps` here?

File: src/components/index.d.ts (hunk 4 above)
Comment (roblevintennis) [reply src/components/index.d.ts]: 🤦‍♂ yup. fixing, thanks Steph

File: src/components/Accordion/Accordion.js (hunk 7 above)
Comment (roblevintennis) [reply src/components/Accordion/Accordion.js]: yeah, or as a default prop in props but I'll just add here 👍

File: src/components/Accordion/useAccordionState.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import { useState } from 'react'
+
+export const useAccordionState = (initialState) => {
```
Comment (stephaniel96) [inline src/components/Accordion/useAccordionState.js]: Just to make sure I'm understanding everything, is `useAccordionState` only meant to be used in the example to show how `onToggle` and `expanded` props work?

File: src/components/Accordion/useAccordionContext.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+export const AccordionContext = createContext({
+  focusRef: {},
+  selected: [null],
+  toggleChildIsTarget: false,
+  expandedAll: [],
+  onToggle: undefined,
+  onNavigation: () => undefined,
+  id: null,
+})
+export const useAccordionContext = () => useContext(AccordionContext)
```
Comment (stephaniel96) [inline src/components/Accordion/useAccordionContext.js]: Most of the documentation on context that I was reading (https://reactjs.org/docs/context.html) described context before hooks were introduced and I was briefly confused. 😂 

This looks great to me!

File: src/components/Accordion/useAccordionState.js (hunk 8 above)
Comment (roblevintennis) [reply src/components/Accordion/useAccordionState.js]: yeah, the consumer can set everything up with something like:
`const { expanded, onToggle } = useAccordionState({ 1: true })`

**Pull request merged**